lib/gnat/services/service_responder.ex

```diff
@@ -2,257 +2,40 @@ defmodule Gnat.Services.ServiceResponder do
-          subject_map: build_subject_map(settings.service_config.endpoints),
-          connection_pid: Map.get(settings, :connection_pid),
-          instance_id: :crypto.strong_rand_bytes(12) |> Base.encode64,
-          subscription: nil,
-          started: DateTime.to_iso8601(DateTime.utc_now())
-        }
-
-        {:ok, subscription} = case get_in(state, [:config, :queue_group]) do
-          nil -> Gnat.sub(state.connection_pid, self(), @subscription_subject)
-          queue_group -> Gnat.sub(state.connection_pid, self(), @subscription_subject, queue_group: queue_group)
-        end
-
-        {:ok, %{ state | subscription: subscription }}
-      end
-
-  end
-
-  @impl true
-  def handle_info({:msg, %{topic: topic, body: _body, reply_to: rt, gnat: gnat}}, state) do
-    tokens = String.split(topic, ".")
-    if length(tokens) >= 2 do
-      operation = Enum.at(tokens, 1)
-      tail = tokens |> Enum.slice(2, length(tokens)-2)
-      case operation do
-        @op_ping -> handle_ping(tail, state, rt, gnat)
-        @op_info -> handle_service_info(tail, state, rt, gnat)
-        @op_stats -> handle_stats(tail, state, rt, gnat)
-      end
-    end
-
-    {:noreply, state}
-  end
-
-
-  @impl GenServer
-  def terminate(:shutdown, state) do
-    Logger.info "#{__MODULE__} starting graceful shutdown"
-    Gnat.unsub(state.connection_pid, state.subscription)
-    Logger.info "#{__MODULE__} finished graceful shutdown"
-  end
-  def terminate(reason, _state) do
-    Logger.error "#{__MODULE__} unexpected shutdown #{inspect reason}"
-  end
-
-  @impl true
-  def handle_cast({:record_request, subject, elapsed_ns}, state) do
-    counters = case :ets.lookup(:endpoint_stats, subject) do
-      [{^subject, counters, _last_error}] ->
-        counters
-      [] ->
-        c = :counters.new(3, [:atomics])
-        :ets.insert(:endpoint_stats, {subject, c, nil})
-        c
-    end
-    :counters.add(counters, @idx_requests, 1)
-    :counters.add(counters, @idx_processing_time, elapsed_ns)
-
-    {:noreply, state}
-  end
-
-  @impl true
-  def handle_cast({:record_error, subject, elapsed_ns, msg}, state) do
-    counters = case :ets.lookup(:endpoint_stats, subject) do
-      [{^subject, counters, _last_error}] ->
-        counters
-      [] ->
-        c = :counters.new(3, [:atomics])
-        :ets.insert(:endpoint_stats, {subject, c, nil})
-        c
-    end
-    :counters.add(counters, @idx_errors, 1)
-    :counters.add(counters, @idx_processing_time, elapsed_ns)
-    :ets.insert(:endpoint_stats, {subject, counters, msg})
-
-    {:noreply, state}
-  end
-
-  @impl true
-  def handle_call({:lookup_endpoint, subject}, _from, state) do
-    res = case Map.get(state.subject_map, subject) do
-      nil ->
-        {nil, nil}
-      {epname, groupname} ->
-        {epname, groupname}
+  def maybe_respond(%{topic: topic} = message, service) do
+    case String.split(topic, ".") do
+      ["$SRV", @op_ping | rest] -> handle_ping(rest, service, message)
```

Comment:
I liked pattern matching to grab the prefix, op and `rest` or `tail` this way rather than using `List.slice`. What do you think?

Reply:
yeah this looks much cleaner